lib/Sema/CSApply.cpp

```diff
@@ -2738,23 +2755,59 @@ namespace {
         // before taking a single element.
         auto baseTy = cs.getType(base);
         if (!toType->hasLValueType() && baseTy->hasLValueType())
-          base = coerceToType(base, baseTy->getRValueType(), cs.getConstraintLocator(base));
+          base = coerceToType(base, baseTy->getRValueType(),
+                              cs.getConstraintLocator(base));
 
         return cs.cacheType(new (cs.getASTContext())
                             TupleElementExpr(base, dotLoc,
                                              selected.choice.getTupleIndex(),
                                              nameLoc.getBaseNameLoc(), toType));
       }
 
-      case OverloadChoiceKind::BaseType: {
+      case OverloadChoiceKind::BaseType:
         return base;
-      }
 
       case OverloadChoiceKind::KeyPathApplication:
         llvm_unreachable("should only happen in a subscript");
+          
+      case OverloadChoiceKind::DynamicMemberLookup: {
+        // Application of a DynamicMemberLookup result turns a member access of
+        // x.foo into x[dynamicMember: "foo"].
+        auto subscriptDecl = cast<SubscriptDecl>(selected.choice.getDecl());
+        auto &ctx = cs.getASTContext();
+        auto loc = nameLoc.getStartLoc();
+        
+        // Build and type check the string literal index value to the specific
+        // string type expected by the subscript.
+        Expr *nameExpr = new (ctx)
+          StringLiteralExpr(selected.choice.getName().getBaseIdentifier().str(),
+                            loc, /*implicit*/true);
+        auto nameLabel = ctx.getIdentifier("dynamicMember");
+        
+        auto stringType = subscriptDecl->getIndices()->get(0)->getTypeLoc();
```

Comment:
If the subscript were generic (or in a generic type), `stringType` could end up being a contextual type (i.e., an archetype) within the subscript itself... which isn't what we want. We really need the interface type to be "opened" so the generic type parameters become type variables, and then simplify `stringType` to get the final result. Here's the kind of wacky case I have in mind:

```
subscript<T: ExpressibleByStringLiteral> (dynamicMember name: T) -> MemberLookup<T> { ... }
```

where the context of the member lookup can affect the type used for the incoming string literal.

Reply:
Got it, thanks again for your help with this case, I really appreciate it.